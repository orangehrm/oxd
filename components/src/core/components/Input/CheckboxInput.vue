<template>
  <div class="oxd-checkbox-wrapper">
    <label :class="{'--disabled': disabled}">
      <template v-if="labelPosition === 'left'">
        {{ optionLabel }}
      </template>
      <input
        type="checkbox"
        @focus="onFocus"
        @blur="onBlur"
        @click="onClick"
        @change="onChange"
        :disabled="disabled"
        v-bind="$attrs"
        v-model="checked"
      />
      <span :class="classes" :style="style" class="oxd-checkbox-input">
        <oxd-icon class="oxd-checkbox-input-icon" :name="checkIcon" />
      </span>
      <template v-if="labelPosition === 'right'">
        {{ optionLabel }}
      </template>
    </label>
  </div>
</template>

<script lang="ts">
import {defineComponent} from 'vue';
import {Position, LABEL_POSITIONS, RIGHT} from './types';
import Icon from '@orangehrm/oxd/core/components/Icon/Icon.vue';

export interface State {
  focused: boolean;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  checkedProxy: any;
}

export default defineComponent({
  name: 'oxd-checkbox-input',
  inheritAttrs: false,
  props: {
    modelValue: {},
    style: {
      type: Object,
    },
    hasError: {
      type: Boolean,
      default: false,
    },
    disabled: {
      type: Boolean,
      default: false,
    },
    checkIcon: {
      type: String,
      default: 'check',
    },
    optionLabel: {
      type: String,
      default: '',
    },
    labelPosition: {
      type: String,
      default: RIGHT,
      validator: function(value: Position) {
        return LABEL_POSITIONS.indexOf(value) !== -1;
      },
    },
  },

  emits: ['update:modelValue', 'blur', 'focus', 'change'],

  components: {
    'oxd-icon': Icon,
  },

  data(): State {
    return {
      focused: false,
      checkedProxy: {},
    };
  },

  computed: {
    classes(): object {
      return {
        'oxd-checkbox-input': true,
        'oxd-checkbox-input--active': !this.focused,
        'oxd-checkbox-input--focus': this.focused,
        'oxd-checkbox-input--error': this.hasError,
        [`--label-${this.labelPosition}`]: true,
      };
    },
    checked: {
      get() {
        return this.modelValue;
      },
      set(value) {
        this.checkedProxy = value;
      },
    },
  },

  methods: {
    onFocus(e: Event) {
      this.focused = true;
      this.$emit('focus', e);
    },
    onBlur(e: Event) {
      this.focused = false;
      this.$emit('blur', e);
    },
    onClick(e: Event) {
      this.focused = false;
      this.$emit('click', e);    
    },
    onChange(e: Event) {
      this.$emit('update:modelValue', this.checkedProxy);
      this.$emit('change', e);
    },
  },
});
</script>

<style src="./checkbox-input.scss" lang="scss" scoped></style>
