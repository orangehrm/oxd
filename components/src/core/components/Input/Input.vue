<template>
<div :class="outerClasses" class="input-outer-wrapper">
  <div v-if="imageIcon" class="input-text-field-icon"> 
    <img :src="imageIcon" />
  </div>
  <oxd-divider v-if="imageIcon" :orientation="'vertical'" :style="{ height:'100%' }"/>
  <input
    :class="classes"
    :style="style"
    :value="modelValue"
    @focus="onFocus"
    @blur="onBlur"
    @input="onInput"
    v-bind="$attrs"
  />
</div>
</template>

<script lang="ts">
import {defineComponent} from 'vue';
import Divider from '@orangehrm/oxd/core/components/Divider/Divider.vue';

export interface State {
  focused: boolean;
}

export default defineComponent({
  name: 'oxd-input',

  components :{
    'oxd-divider' : Divider
  },

  props: {
    // https://v3.vuejs.org/guide/migration/v-model.html#overview
    modelValue: {},
    style: {
      type: Object,
    },
    hasError: {
      type: Boolean,
      default: false,
    },
    imageIcon: {
      type: String
    }
  },

  data(): State {
    return {
      focused: false,
    };
  },

  computed: {
    classes(): object {
      const isIconExists=this.imageIcon ? '--with-icon' : '';
      return {
        [`oxd-input${isIconExists}`]: true,
        [`oxd-input${isIconExists}--active`]: !this.focused,
        [`oxd-input${isIconExists}--focus`] : this.focused,
        [`oxd-input${isIconExists}--error`] : this.hasError
      };
    },
    outerClasses(): object {
      return this.imageIcon ? {
        'input-outer' : true,
        'input-outer--active': !this.focused,
        'input-outer--focus': this.focused,
        'input-outer--error': this.hasError,
      } : {'input-outer' : false};
    },
  },

  methods: {
    onFocus() {
      this.focused = true;
    },
    onBlur() {
      this.focused = false;
    },
    onInput(e: Event) {
      e.preventDefault();
      this.$emit('update:modelValue', (e.target as HTMLInputElement).value);
      this.$emit('input', e);
    },
  },
});
</script>

<style src="./input.scss" lang="scss" scoped></style>
