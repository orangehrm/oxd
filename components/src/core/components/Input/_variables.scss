// This file is part of OrangeHRM Inc
//
// Copyright (C) 2020 onwards OrangeHRM Inc
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see  http://www.gnu.org/licenses

$oxd-input-min-height: 45px !default;
$oxd-input-border--active: $oxd-input-control-border--active !default;
$oxd-input-border--focus: $oxd-input-control-border--focus !default;
$oxd-input-border--error: $oxd-input-control-border--error !default;
$oxd-input-box-shadow--focus: $oxd-input-control-box-shadow--focus !default;
$oxd-input-box-shadow--error: $oxd-input-control-box-shadow--error !default;
$oxd-input-background--disabled: darken(
  $oxd-background-white-shadow-color,
  3%
) !default;

$oxd-file-input-horizontal-margin: 0.25rem !default;
$oxd-file-input-background-color: $oxd-white-color !default;
$oxd-file-button-background-color: $oxd-background-light-gray-color !default;
$oxd-file-button-color: $oxd-interface-gray-darken-1-color !default;
$oxd-file-button-font-size: 0.875rem;
$oxd-file-button-font-weight: 600;
$oxd-file-button-padding-horizontal: 1.5rem;
$oxd-file-button-color--disabled: $oxd-interface-gray-lighten-1-color !default;
$oxd-file-button-background-color--disabled: darken(
  $oxd-white-color,
  1%
) !default;
$oxd-file-button-height: 35px !default;
$oxd-file-button-width: 35px !default;

$oxd-checkbox-label-color: $oxd-input-control-font-color !default;
$oxd-checkbox-label-color--disabled: $oxd-interface-gray-color !default;
$oxd-checkbox-check-font-size: 1rem !default;
$oxd-checkbox-height: 1rem !default;
$oxd-checkbox-width: 1rem !default;
$oxd-checkbox-border: 1px solid $oxd-interface-gray-lighten-1-color !default;
$oxd-checkbox-border-radius: 0.35rem !default;
$oxd-checkbox-label-margin: 0.5em !default;
$oxd-checkbox-checked-background-color: $oxd-primary-one-color !default;
$oxd-checkbox-disabled-background-color: $oxd-interface-gray-lighten-2-color !default;

$oxd-switch-label-color: $oxd-input-control-font-color !default;
$oxd-switch-label-color--disabled: $oxd-interface-gray-color !default;
$oxd-switch-width: 2rem !default;
$oxd-switch-height: calc(#{$oxd-switch-width} / 2);
$oxd-switch-radius: calc(#{$oxd-switch-height} / 2);
$oxd-switch-selector-height: calc(#{$oxd-switch-height} - 2px);
$oxd-switch-selector-width: calc(#{$oxd-switch-height} - 2px);
$oxd-switch-selector-radius: calc(#{$oxd-switch-selector-height} / 2);
$oxd-switch-selector-color: $oxd-white-color !default;
$oxd-switch-label-margin: 0.5em;
$oxd-switch-off-background-color: $oxd-interface-gray-lighten-2-color !default;
$oxd-switch-on-background-color: $oxd-primary-one-color !default;
$oxd-switch-on-disabled-background-color: rgba(
  $oxd-primary-one-color,
  0.5
) !default;
$oxd-switch-off-disabled-background-color: $oxd-interface-gray-lighten-2-color !default;
$oxd-switch-disabled-background-color: $oxd-interface-gray-lighten-2-color !default;

$oxd-radio-label-color: $oxd-input-control-font-color !default;
$oxd-radio-label-color--disabled: $oxd-interface-gray-color !default;
$oxd-radio-height: 1rem !default;
$oxd-radio-width: 1rem !default;
$oxd-radio-label-margin: 0.5em !default;
$oxd-radio-checked-color: $oxd-primary-one-color !default;
$oxd-radio-checked-disabled-color: lighten(
  $oxd-radio-checked-color,
  30%
) !default;

$oxd-dropdown-text-padding: 0 0.25rem 0 0.5rem !default;
$oxd-dropdown-animation-time: 300ms !default;
$oxd-dropdown-max-height: 11.5rem !default;
$oxd-dropdown-loader-height: 1rem !default;
$oxd-dropdown-dropdown-z-index: $oxd-dropdown-z-index !default;
$oxd-dropdown-dropdown-box-shadow: 0px 5px 15px -8px rgba(0, 0, 0, 0.15) !default;
$oxd-dropdown-dropdown-padding: 4px !default;
$oxd-dropdown-dropdown-border: 1px solid $oxd-interface-gray-lighten-2-color !default;
$oxd-dropdown-dropdown-border-radius: 0.65rem !default;
$oxd-dropdown-dropdown-background: $oxd-white-color !default;
$oxd-dropdown-dropdown-background--disabled: darken(
  $oxd-background-white-shadow-color,
  3%
) !default;
$oxd-dropdown-dropdown-margin: 5px !default;
$oxd-dropdown-dropdown-color: $oxd-interface-gray-darken-1-color !default;
$oxd-dropdown-dropdown-slot-color: $oxd-interface-gray-color !default;
$oxd-dropdown-dropdown-max-height: 224px !default;

$oxd-dropdown-option-font-size: $oxd-input-control-font-size !default;
$oxd-dropdown-option-font-family: $oxd-input-control-font-family !default;
$oxd-dropdown-option-font-weight: $oxd-input-control-font-weight !default;
$oxd-dropdown-option-padding: 4px !default;
$oxd-dropdown-option-min-height: 27px !default;
$oxd-dropdown-option-line-height: normal !default;
$oxd-dropdown-option-border-bottom: 1px solid #fbfbfc !default;
$oxd-dropdown-option-border-radius: 0.65rem !default;
$oxd-dropdown-option-color: $oxd-interface-gray-darken-1-color !default;
$oxd-dropdown-option-color--disabled: $oxd-interface-gray-lighten-1-color !default;
$oxd-dropdown-option-color--selected: $oxd-interface-gray-lighten-1-color !default;
$oxd-dropdown-option-background: transparent !default;
$oxd-dropdown-option-background--focused: #f6f7f9 !default;

$oxd-dropdown-chip-margin: 5px !default;
$oxd-dropdown-chip-background-color: #f6f7f9 !default;
$oxd-dropdown-chip-clear-background-color: $oxd-interface-gray-lighten-1-color !default;
$oxd-dropdown-chip-clear-font-color: $oxd-white-color !default;
$oxd-dropdown-chip-border-radius: 50% !default;
$oxd-dropdown-chip-area-padding: 0.5rem 0 !default;

$oxd-dropdown-arrow-background-color: $oxd-background-light-gray-color !default;
$oxd-dropdown-arrow-background-color--disabled: $oxd-background-light-gray-color !default;
$oxd-dropdown-arrow-background-color--readonly: darken(
  $oxd-white-color,
  1%
) !default;
$oxd-dropdown-arrow-color--readonly: $oxd-interface-gray-lighten-1-color !default;
$oxd-dropdown-arrow-width: 35px !default;
$oxd-dropdown-arrow-height: 35px !default;
$oxd-dropdown-arrow-border-radius: 0.65rem !default;

$oxd-time-picker-period-label-width: 60px !default;
$oxd-time-picker-period-label-padding: 0.25rem 0rem !default;
$oxd-time-picker-period-horizontal-padding: 16px !default;
$oxd-time-picker-period-vertical-padding: 8px !default;
$oxd-time-picker-period-font-size: 14px !default;
$oxd-time-picker-period-font-weight: 600 !default;
$oxd-time-picker-period-font-family: $oxd-font-family !default;
$oxd-time-picker-period-font-color: $oxd-white-color !default;
$oxd-time-picker-period-border-radius: 0.5rem !default;
$oxd-time-picker-dropdown-margin: 1.5rem 1rem !default;
$oxd-time-picker-dropdown-width: 250px !default;
$oxd-time-picker-dropdown-input-width: 60px !default;
$oxd-time-picker-dropdown-seperator-padding: 0.5rem !default;

$oxd-date-input-links-margin: 0.5rem !default;
$oxd-date-input-link-padding: 5px 10px !default;
$oxd-date-input-link-border-radius: 0.75rem !default;
$oxd-date-input-link-font-family: $oxd-input-control-font-family !default;
$oxd-date-input-link-font-color: $oxd-input-control-font-color !default;
$oxd-date-input-link-font-weight: 500 !default;
$oxd-date-input-link-font-size: 14px !default;
$oxd-date-input-link-highlight-color: $oxd-primary-one-color !default;
$oxd-date-input-link-background: $oxd-background-light-gray-color !default;
