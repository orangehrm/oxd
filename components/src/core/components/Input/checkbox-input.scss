@import '../../../styles';
@import 'variables';

.oxd-checkbox-wrapper {
  & label {
    cursor: pointer;
    display: flex;
    align-content: center;
    font-family: $oxd-input-control-font-family;
    color: $oxd-checkbox-label-color;
    font-size: $oxd-input-control-font-size;
    font-weight: $oxd-input-control-font-weight;
    width: fit-content;

    &:after {
      content: '';
      display: table;
      clear: both;
    }

    & input[type='checkbox'] {
      opacity: 0;
      position: absolute;
      + .oxd-checkbox-input {
        background-color: $oxd-white-color;
        .oxd-checkbox-input-icon {
          opacity: 0;
        }
      }

      &:checked + .oxd-checkbox-input {
        background-color: $oxd-checkbox-checked-background-color;
        .oxd-checkbox-input-icon {
          opacity: 1;
        }
      }

      &:checked:disabled + .oxd-checkbox-input {
        background-color: $oxd-checkbox-disabled-background-color;
        .oxd-checkbox-input-icon {
          opacity: 1;
        }
      }

      &:disabled + .oxd-checkbox-input {
        background-color: $oxd-checkbox-disabled-background-color;
        .oxd-checkbox-input-icon {
          opacity: 0;
        }
      }
    }

    &.--disabled {
      color: $oxd-checkbox-label-color--disabled;
      cursor: unset;
    }
  }
}

.oxd-checkbox-input {
  position: relative;
  display: flex;
  border-radius: $oxd-checkbox-border-radius;
  width: $oxd-checkbox-width;
  height: $oxd-checkbox-height;

  &.--label-left {
    margin-left: $oxd-checkbox-label-margin;
  }

  &.--label-right {
    margin-right: $oxd-checkbox-label-margin;
  }

  &--active {
    border: $oxd-checkbox-border;
  }

  &--focus {
    border: $oxd-input-border--focus;
    box-shadow: $oxd-input-box-shadow--focus;
    outline: 0;
  }

  &--error {
    border: $oxd-input-border--error;
    box-shadow: $oxd-input-box-shadow--error;
  }

  &-icon {
    color: $oxd-white-color;
    font-size: $oxd-checkbox-check-font-size;
    position: absolute;
    line-height: 1.5;
    align-self: center;
  }
}
