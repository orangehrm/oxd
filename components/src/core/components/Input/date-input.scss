@import '../../../styles';
@import 'variables';

.oxd-date-wrapper {
  @include oxd-input-control-size;
  position: relative;
  &:focus {
    outline: none;
  }
  & > *:focus {
    outline: none;
  }
}

.oxd-date-input-calendar {
  &.--positon-top > .oxd-calendar-wrapper {
    right: 0;
    width: 88%;
    position: absolute;
    bottom: 100%;
    margin-bottom: calc(50px + $oxd-dropdown-dropdown-margin);
  }
  &.--positon-bottom > .oxd-calendar-wrapper {
    right: 0;
    width: 271px;
    position: absolute;
  }
}

.oxd-date-input {
  display: flex;
  align-items: center;
  &-icon-wrapper {
    right: 0.25rem;
    display: flex;
    cursor: pointer;
    position: absolute;
    align-items: center;
    justify-content: center;
    width: $oxd-dropdown-arrow-width;
    height: $oxd-dropdown-arrow-height;
    border-radius: $oxd-dropdown-arrow-border-radius;
    background-color: $oxd-dropdown-arrow-background-color;
    color: $oxd-file-button-color;
    &.--disabled,
    &.--readonly {
      cursor: unset;
      background-color: $oxd-dropdown-arrow-background-color--readonly;
      color: $oxd-dropdown-arrow-color--readonly;
    }
    &:focus {
      background-color: $oxd-interface-gray-lighten-2-color;
      border: $oxd-input-border--focus;
      box-shadow: $oxd-input-box-shadow--focus;
      outline: 0;
    }
  }
  ::v-deep(.oxd-input) {
    text-overflow: ellipsis;
    padding-right: 4rem;
  }
  &-calendar {
    top: 50px;
    right: 0;
    position: absolute;
    z-index: $oxd-date-z-index;
    width: 250px;
    @include oxd-respond-to('sm') {
      width: 300px;
    }
  }
  &-links {
    display: flex;
    justify-content: space-between;
    align-items: center;
    margin-top: $oxd-date-input-links-margin;
  }
  &-link {
    cursor: pointer;
    padding: $oxd-date-input-link-padding;
    border-radius: $oxd-date-input-link-border-radius;
    font-family: $oxd-input-control-font-family;
    color: $oxd-input-control-font-color;
    font-weight: $oxd-date-input-link-font-weight;
    font-size: $oxd-date-input-link-font-size;
    &.--today,
    &.--close {
      color: $oxd-date-input-link-highlight-color;
    }
    &:hover,
    &:focus-visible {
      font-weight: $oxd-date-input-link-font-weight-hover;
    }
  }
}

.transition-fade-down-enter-from,
.transition-fade-down-leave-to {
  transform: translateY(-20px);
  opacity: 0;
}

.transition-fade-down-enter-active,
.transition-fade-down-leave-active {
  transition: transform $oxd-dropdown-animation-time
      cubic-bezier(0.53, 2, 0.36, 0.85),
    opacity $oxd-dropdown-animation-time cubic-bezier(0.53, 2, 0.36, 0.85);
}

::v-deep(.oxd-loading-spinner) {
  width: $oxd-dropdown-loader-height;
  height: $oxd-dropdown-loader-height;
}
