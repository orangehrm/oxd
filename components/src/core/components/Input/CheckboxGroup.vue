<script lang="ts">
import {defineComponent, h} from 'vue';
import CheckboxInput from '@orangehrm/oxd/core/components/Input/CheckboxInput.vue';
import InputGroup from '@orangehrm/oxd/core/components/InputField/InputGroup.vue';
import useTranslate from '../../../composables/useTranslate';

export interface State {
  focused: boolean;
}

export interface Options {
  id: number;
  label: string;
  disabled?: boolean;
  style?: Record<string, any>;
}

export default defineComponent({
  name: 'oxd-checkbox-group',
  components: {
    CheckboxInput,
    InputGroup,
  },

  emits: ['update:modelValue', 'blur', 'focus', 'change'],

  props: {
    id: {
      type: String,
      default: 'check-box-group-id',
    },
    class: {
      type: String,
      default: 'check-box-column',
    },
    style: {
      type: Object,
    },
    disabled: {
      type: Boolean,
    },
    options: {
      type: Array,
    },
    modelValue: {
      type: Array,
      default: () => [],
    },
  },

  render() {
    const inputId = this.id == '' ? 'check-box-group-id' : this.id;
    const inputClass = this.class == '' ? 'check-box-column' : this.class;
    const {$t} = useTranslate();
    return h(
      InputGroup,
      {
        classes: {
          wrapper: inputClass,
        },
        id: inputId,
        style: this.style,
      },
      [
        this.options.map((option: Options, index: number) => {
          return h(CheckboxInput, {
            id: inputId + '_' + option.id,
            value: option.id,
            style: option.style,
            autofocus: this.$attrs.autofocus && index === 0 ? true : false,
            optionLabel: $t(option.label),
            modelValue: this.modelValue,
            disabled: this.disabled ? 'true' : option.disabled,
            onFocus: () => {
              this.$emit('focus', true);
            },
            onBlur: () => {
              this.$emit('blur', true);
            },
            onClick: () => {
              this.$emit('click', true);
            },
            onChange: () => {
              let idArray: number[];
              if (this.modelValue.includes(option.id)) {
                idArray = [...this.modelValue].filter(
                  value => value !== option.id,
                );
              } else {
                idArray = [...this.modelValue, option.id];
              }
              this.$emit('update:modelValue', idArray.sort());
              this.$emit('change', true);
            },
          });
        }),
      ],
    );
  },
});
</script>

<style lang="scss" scoped>
:deep(.check-box-row) {
  display: flex;
  gap: 2rem;
}

:deep(.check-box-column) {
  display: flex;
  flex-direction: column;
  gap: 1rem;
}
</style>
