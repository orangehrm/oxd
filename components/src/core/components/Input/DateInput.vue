<template>
  <div class="oxd-date-wrapper">
    <div
      class="oxd-date-input"
      :tooltip="isLengthyDate ? displayDate : null"
      flow="bottom"
    >
      <oxd-input
        :hasError="hasError"
        :disabled="disabled"
        :readonly="readonly"
        :value="displayDate"
        :placeholder="placeholder"
        v-bind="defaultAttrs"
        v-click-outside="onClickTextOutside"
        ref="oxdInput"
        @update:modelValue="onDateTyped"
        @blur="onBlur"
      />
      <div
        class="oxd-date-input-icon-wrapper"
        tabindex="0"
        :class="dateIconAttributeClasses"
        v-if="!disabled"
        ref="oxdIcon"
        @click="toggleDropdown"
        @keyup.enter.prevent.stop="toggleDropdown"
      >
        <oxd-icon :class="dateIconClasses" name="oxd-calendar" size="small" />
      </div>
    </div>
    <transition name="transition-fade-down">
      <div
        v-click-outside="onClickOutside"
        v-if="open"
        class="oxd-date-input-calendar"
        @keyup.esc="closeDropdown"
        v-dropdown-direction
      >
        <oxd-calendar
          @update:modelValue="onDateSelected"
          @selectMonth="selectMonth"
          @selectYear="selectYear"
          @mousedown.prevent
          v-model="dateSelected"
          :locale="locale"
          :firstDayOfWeek="firstDayOfWeek"
          :years="years"
          :monthFormat="monthFormat"
          :months="months"
          :dayFormat="dayFormat"
          :days="days"
          :dayAttributes="dayAttributes"
          :events="events"
          v-focus-trap
        >
          <div class="oxd-date-input-links">
            <div
              @keyup.enter="onClickToday"
              @click="onClickToday"
              class="oxd-date-input-link --today"
              tabindex="0"
            >
              {{ $vt('Today') }}
            </div>
            <div
              @keyup.enter="onClickClear"
              @click="onClickClear"
              class="oxd-date-input-link --clear"
              tabindex="0"
            >
              {{ $vt('Clear') }}
            </div>
            <div
              @keyup.enter="closeDropdown"
              @click="closeDropdown"
              class="oxd-date-input-link --close"
              tabindex="0"
            >
              {{ $vt('Close') }}
            </div>
          </div>
        </oxd-calendar>
      </div>
    </transition>
  </div>
</template>

<script lang="ts">
import {enUS} from 'date-fns/locale';
import {defineComponent, PropType} from 'vue';
import {formatDate, parseDate, freshDate, getYear} from '../../../utils/date';
import Icon from '@orangehrm/oxd/core/components/Icon/Icon.vue';
import Input from '@orangehrm/oxd/core/components/Input/Input.vue';
import Calendar from '@orangehrm/oxd/core/components/Calendar/Calendar.vue';
import clickOutsideDirective from '../../../directives/click-outside';
import dropdownDirectionDirective from '../../../directives/dropdown-direction';
import focusTrapDirective from '../../../directives/focus-trap';
import translateMixin from '../../../mixins/translate';
import {LENGTHY_DATE_FORMATS} from '../Calendar/types';
import {CalendarDayAttributes, CalendarEvent} from '../Calendar/types';

export default defineComponent({
  name: 'oxd-date-input',
  inheritAttrs: false,

  components: {
    'oxd-icon': Icon,
    'oxd-input': Input,
    'oxd-calendar': Calendar,
  },

  mixins: [translateMixin],

  directives: {
    'click-outside': clickOutsideDirective,
    'dropdown-direction': dropdownDirectionDirective,
    'focus-trap': focusTrapDirective,
  },

  props: {
    modelValue: {
      type: String,
      default: '',
    },
    hasError: {
      type: Boolean,
      default: false,
    },
    disabled: {
      type: Boolean,
      default: false,
    },
    readonly: {
      type: Boolean,
      default: false,
    },
    placeholder: {
      type: String,
      required: false,
    },
    ioformat: {
      type: String,
      default: 'yyyy-MM-dd',
    },
    displayFormat: {
      type: String,
      default: null,
    },
    locale: {
      type: Object as PropType<Locale>,
      default: enUS,
    },
    firstDayOfWeek: {
      type: Number,
      default: 0, // 0 | 1 | 2 | 3 | 4 | 5 | 6 => 0 represents Sunday
    },
    years: {
      type: Array,
      default: () => {
        return Array.from(
          {length: getYear(new Date()) - 1969},
          (_, i) => 1970 + i,
        );
      },
    },
    monthFormat: {
      type: String,
      default: 'wide',
    },
    months: {
      type: Array,
      default: () => [],
    },
    dayFormat: {
      type: String,
      default: 'narrow',
    },
    days: {
      type: Array,
      default: () => [],
    },
    dayAttributes: {
      type: Array as PropType<CalendarDayAttributes[]>,
      default: () => [],
    },
    events: {
      type: Array as PropType<CalendarEvent[]>,
      default: () => [],
    },
  },

  data() {
    return {
      open: false,
      dateTyped: null,
    };
  },

  methods: {
    onBlur(e: Event) {
      if (this.dateTyped !== null) {
        this.dateSelected = parseDate(this.dateTyped, this.format, {
          locale: this.locale,
        });
      }
      e.stopImmediatePropagation();
      this.$emit('blur');
    },
    onDateTyped(value: string) {
      this.dateTyped = value;
    },
    onDateSelected() {
      const oxdDatePicker = this.$refs.oxdInput;
      oxdDatePicker.focused = true;
      this.$nextTick(() => {
        const oxdDateInputTriggerBtn = this.$refs.oxdIcon;
        oxdDateInputTriggerBtn.blur();
      });
      this.closeDropdown();
    },
    toggleDropdown() {
      if (!this.disabled && !this.readonly) {
        if (!this.open) {
          this.openDropdown();
        } else {
          this.closeDropdown();
        }
      }
    },
    openDropdown() {
      this.open = true;
      this.$emit('dateselect:opened');
    },
    closeDropdown($e: KeyboardEvent | null) {
      if ($e && $e.key === 'Escape' && this.open) $e.stopPropagation();
      this.open = false;
      this.$refs.oxdIcon.focus();
      this.$emit('dateselect:closed');
    },
    onClickOutside() {
      this.open = false;
      this.$emit('dateselect:closed');
    },
    onClickTextOutside() {
      const oxdDatePicker = this.$refs.oxdInput;
      oxdDatePicker.focused = false;
    },
    onClickToday() {
      this.dateSelected = freshDate();
      this.open = false;
      this.$refs.oxdIcon.focus();
    },
    onClickClear() {
      this.dateTyped = '';
      this.dateSelected = null;
      this.open = false;
      this.$refs.oxdIcon.focus();
    },
    selectMonth($e: Event) {
      this.$emit('selectMonth', $e);
    },
    selectYear($e: Event) {
      this.$emit('selectYear', $e);
    },
  },

  computed: {
    defaultAttrs() {
      const notAllowed = ['onUpdate:modelValue'];
      return Object.keys(this.$attrs)
        .filter(key => !notAllowed.includes(key))
        .reduce((obj, key) => {
          obj[key] = this.$attrs[key];
          return obj;
        }, {});
    },
    isLengthyDate() {
      return LENGTHY_DATE_FORMATS.indexOf(this.format) > -1;
    },
    dateSelected: {
      get() {
        return parseDate(this.modelValue, this.ioformat);
      },
      set(value) {
        this.$emit(
          'update:modelValue',
          formatDate(value, this.ioformat) ||
            (this.dateTyped ? this.dateTyped : null),
        );
      },
    },
    displayDate(): string {
      return (
        formatDate(this.dateSelected, this.format, {locale: this.locale}) ||
        this.dateTyped
      );
    },
    dateIconClasses(): object {
      return {
        'justify-center': true,
        'oxd-date-input-icon': true,
        '--disabled': this.disabled,
        '--readonly': this.readonly,
      };
    },
    dateIconAttributeClasses(): object {
      return {
        '--disabled': this.disabled,
        '--readonly': this.readonly,
      };
    },
    format(): string {
      return this.displayFormat && this.displayFormat.trim()
        ? this.displayFormat
        : this.ioformat;
    },
  },
});
</script>

<style src="./date-input.scss" lang="scss" scoped></style>
