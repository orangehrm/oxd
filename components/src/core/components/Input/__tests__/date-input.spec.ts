import {mount} from '@vue/test-utils';
import DateInput from '@orangehrm/oxd/core/components/Input/DateInput.vue';
import Input from '@orangehrm/oxd/core/components/Input/Input.vue';
import Icon from '@orangehrm/oxd/core/components/Icon/Icon.vue';
import {formatDate, freshDate} from '../../../../utils/date';

describe('DateInput.vue', () => {
  it('renders OXD Date Input', () => {
    const wrapper = mount(DateInput, {});
    expect(wrapper.html()).toMatchSnapshot();
  });
  it('should open datepicker on click', async () => {
    const wrapper = mount(DateInput, {});
    wrapper.findComponent(Icon).trigger('click');
    await wrapper.vm.$nextTick();
    expect(wrapper.vm.open).toBeTruthy();
    expect(wrapper.emitted('dateselect:opened')).toBeTruthy();
    expect(wrapper.find('.oxd-calendar-wrapper').exists()).toBeTruthy();
  });
  it('should not open datepicker on click, if disabled', async () => {
    const wrapper = mount(DateInput, {
      props: {disabled: true},
    });
    wrapper.findComponent(Icon).trigger('click');
    await wrapper.vm.$nextTick();
    expect(wrapper.vm.open).toBeFalsy();
    expect(wrapper.emitted('dateselect:opened')).toBeFalsy();
    expect(wrapper.find('.oxd-calendar-wrapper').exists()).toBeFalsy();
  });
  it('should accept valid input', async () => {
    const wrapper = mount(DateInput, {});
    const input = wrapper.findComponent(Input);
    (input.element as HTMLInputElement).value = '2021-07-30';
    await input.trigger('input');
    await input.trigger('blur');
    await wrapper.vm.$nextTick();
    expect(wrapper.emitted('update:modelValue')).toEqual([['2021-07-30']]);
  });
  it('should return today date onclick today', async () => {
    const wrapper = mount(DateInput, {});
    await wrapper.findComponent(Icon).trigger('click');
    await wrapper.vm.$nextTick();
    await wrapper.find('.oxd-date-input-link.--today').trigger('click');
    const dateExpected = formatDate(freshDate(), 'yyyy-MM-dd');
    expect(wrapper.emitted('update:modelValue')).toEqual([[dateExpected]]);
  });
  it('should clear date onclick clear', async () => {
    const wrapper = mount(DateInput, {});
    await wrapper.findComponent(Icon).trigger('click');
    await wrapper.vm.$nextTick();
    await wrapper.find('.oxd-date-input-link.--clear').trigger('click');
    expect(wrapper.emitted('update:modelValue')).toEqual([[null]]);
  });
  it('should close datepicker onclick close', async () => {
    const wrapper = mount(DateInput, {});
    await wrapper.findComponent(Icon).trigger('click');
    await wrapper.vm.$nextTick();
    await wrapper.find('.oxd-date-input-link.--close').trigger('click');
    expect(wrapper.emitted('dateselect:closed'))?.toBeTruthy();
  });
});
