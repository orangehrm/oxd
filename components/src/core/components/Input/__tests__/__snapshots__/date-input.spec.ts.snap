// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateInput.vue renders OXD Date Input 1`] = `
<div class="oxd-date-wrapper">
  <div class="oxd-date-input"><input class="oxd-input oxd-input--active"><i class="oxd-icon bi-calendar oxd-date-input-icon"></i></div>
  <transition-stub>
    <!--v-if-->
  </transition-stub>
</div>
`;
